lib/collection/src/shards/resharding/driver.rs

```diff
@@ -514,16 +530,92 @@ fn stage_commit_hashring() -> CollectionResult<()> {
-    todo!()
+fn completed_propagate_deletes(state: &PersistedState) -> bool {
+    let state_read = state.read();
+    state_read.all_peers_reached(Stage::S5_PropagateDeletesEnd)
+        && state_read.shards_to_delete().next().is_none()
 }
 
 /// Stage 5: commit new hashring
 ///
 /// Do delete migrated points from their old shards.
-fn stage_propagate_deletes() -> CollectionResult<()> {
+// TODO(resharding): this is a naive implementation, delete by hashring filter directly!
+async fn stage_propagate_deletes(
+    reshard_key: &ReshardKey,
+    state: &PersistedState,
+    shard_holder: Arc<LockedShardHolder>,
+) -> CollectionResult<()> {
     log::debug!("Resharding stage: propagate deletes");
-    todo!()
+
+    state.write(|data| {
+        data.bump_all_peers_to(Stage::S5_PropagateDeletesStart);
+    })?;
+
+    let hashring = {
+        let shard_holder = shard_holder.read().await;
+        let shard_key = shard_holder
+            .get_shard_id_to_key_mapping()
+            .get(&reshard_key.shard_id)
+            .cloned();
+        shard_holder.rings.get(&shard_key).cloned().ok_or_else(|| {
+            CollectionError::service_error(format!(
+                "Cannot delete migrated points while resharding shard {}, failed to get shard hash ring",
+                reshard_key.shard_id,
+            ))
+        })?
+    };
+
+    while let Some(source_shard_id) = block_in_place(|| state.read().shards_to_delete().next()) {
+        let mut offset = None;
+
+        while {
+            let shard_holder = shard_holder.read().await;
+            let replica_set = shard_holder.get_shard(&source_shard_id).ok_or_else(|| {
+                CollectionError::service_error(format!(
+                    "Shard {source_shard_id} not found in the shard holder for resharding",
+                ))
+            })?;
+
+            // Take batch of points, if full, pop the last entry as next batch offset
+            let mut batch = replica_set
+                .scroll_by(
+                    offset,
+                    DELETE_BATCH_SIZE + 1,
+                    &false.into(),
+                    &false.into(),
+                    // TODO(resharding): directly apply hash ring filter here
+                    None,
+                    None,
+                    false,
+                    None,
+                )
+                .await?;
+            offset = (batch.len() >= DELETE_BATCH_SIZE).then(|| batch.pop().unwrap().id);
+
+            let ids = batch
+                .into_iter()
+                .filter(|record| hashring.has_moved(&record.id))
+                .map(|record| record.id)
+                .collect::<Vec<_>>();
+            let operation =
+                CollectionUpdateOperations::PointOperation(PointOperations::DeletePoints { ids });
+            replica_set
+                .update_with_consistency(operation, offset.is_some(), WriteOrdering::Weak)
+                .await?;
+
+            offset.is_some()
+        } {}
```

Comment:
This loop is cursed. 😂

Why not either:
```rust
let mut spin = true;

while spin {
  // ...

  spin = offset.is_some();
}
```

Or:
```rust
loop {
  // ...

  if offset.is_none() {
    break;
  }
}
```

Reply:
It's a do-while loop, this is how Rust imagined it :smile: 

Fine with both your alternatives though!

Reply:
Fixed.